server/src/main/java/org/elasticsearch/health/metadata/HealthMetadataService.java

```diff
@@ -124,39 +132,32 @@ private void enable(boolean enabled) {
             resetHealthMetadata("health-node-enabled");
         } else {
             clusterService.removeListener(clusterStateListener);
-            readyToPublish = false;
+            readyToPublishAsMaster = false;
         }
     }
 
     private void updateOnClusterStateChange(ClusterChangedEvent event) {
         final boolean wasMaster = event.previousState().nodes().isLocalNodeElectedMaster();
         isMaster = event.localNodeMaster();
         if (isMaster && wasMaster == false) {
-            readyToPublish = true;
+            readyToPublishAsMaster = true;
         } else if (isMaster == false) {
-            readyToPublish = false;
+            readyToPublishAsMaster = false;
         }
+
         // Wait until every node in the cluster is upgraded to 8.5.0 or later
         if (event.state().nodesIfRecovered().getMinNodeVersion().onOrAfter(Version.V_8_5_0)) {
-            if (readyToPublish) {
+            if (readyToPublishAsMaster) {
                 resetHealthMetadata("health-metadata-update-master-election");
-                readyToPublish = false;
-            }
-        }
-    }
-
-    private void updateOnSettingsUpdated(UpsertHealthMetadataTask task) {
-        // We do not use the cluster state to check if this is the master node because the cluster state might not have been initialized
-        if (isMaster && enabled) {
-            ClusterState clusterState = clusterService.state();
-            if (clusterState.nodesIfRecovered().getMinNodeVersion().onOrAfter(Version.V_8_5_0)) {
-                taskQueue.submitTask("health-metadata-update", task, null);
+                readyToPublishAsMaster = false;
+            } else if (isMaster && event.state().blocks().hasGlobalBlock(GatewayService.STATE_NOT_RECOVERED_BLOCK) == false) {
+                taskQueue.submitTask("health-metadata-update", () -> this.localHealthMetadata, null);
```

Comment:
If I understand this code correctly we will submit one cluster state update with the local health metadata for ... every cluster state that this master node receives (even if the local health metadata has not changed since we last published it). I believe we shouldn't send more cluster state updates than needed here to not add more things for the master node to process (note that after it receives the cluster state update request here, it will drop the request as there isn't anything to update) 

I wonder if we should take a step back here and sketch how we think this health metadata upsert functionality should work theoretically (and then translate it into code) - we haven't really had this use case before in Elasticsearch (reacting to a cluster setting by publishing a cluster state update, and doing an initialise operation on master failover).

For e.g. I think we shouldn't do anything if `event.state().blocks().hasGlobalBlock(GatewayService.STATE_NOT_RECOVERED_BLOCK)` is `true`

I think we can already write down a few guiding invariants e.g. 

1. The first line in `updateOnClusterStateChange` should be:

```
if (event.state().blocks().hasGlobalBlock(GatewayService.STATE_NOT_RECOVERED_BLOCK) ) {
  return;
}
```
(we shouldn't make decisions and act on a partially recovered cluster state - the cluster settings updates will be received by the settings listeners and they'll update the local state)

2. we should only publish the local health metadata only if it has changed from what we published before 
3. the cluster state update we publish on master failover should not override a setting update

Reply:
> 1. The first line in updateOnClusterStateChange should be:
if (event.state().blocks().hasGlobalBlock(GatewayService.STATE_NOT_RECOVERED_BLOCK) ) {
  return;
}
(we shouldn't make decisions and act on a partially recovered cluster state - the cluster settings updates will be received by the settings listeners and they'll update the local state)

100% agree

> 2. we should only publish the local health metadata only if it has changed from what we published before

I agree but I don't know how to do it without having a `previousHealthMetadata` object, as a baseline to check against. In previous commits, I did that but changed it by @gmarouli's recommendation. Mary, what do you think of this? Any other idea of maintaining only 1 object in memory? or maybe a flag indicating that _something_ has changed? 

> 3. the cluster state update we publish on master failover should not override a setting update

We're doing this with the current code, as long as each node properly reacts to setting changes.

Thinking a bit more, I think we've no tests checking a setting update in the cluster. will check and write a test for that.

Reply:
>I agree but I don't know how to do it without having a previousHealthMetadata object

Could we compare it with whatever's already in the cluster state? ( note that we wouldn't want to reset the settings _back_ to the default values - the defaults might be different to what's in the cluster state however, by the time the cluster state is fully recovered we should've received the latest cluster settings ) 

e.g.
```
if (localHealth.equals(clusterState.custom(HealthMetadata.TYPE) == false) {
  // our local is different than what's in the cluster state 
  // fire cluster state update
}
```

Reply:
I completely agree to compare with the health metadata that is in the cluster state. 

I see my initial suggestion suggestion was not clear. I thought we can skip the comparison if we would only publish the health metadata when we witness a new master or a setting update. Not as a reaction to every cluster update.

But the code looks simpler the way you approached it, so if we compare the local view of health metadata with what we retrieve from the cluster state, we should be fine.